004 Start work on governance tidyup (#3)
tidyverse/tidyups: 004-governance.Rmd

-   Providing moral support (a "thank you" goes a long way).

### Contributors

Users who continue to engage with the project and its community will often find themselves becoming more and more involved.
Such users may then go on to become **contributors** by interacting with the project on GitHub.
Contributors:

-   Report bugs and suggest improvements by creating new issues.

-   Improving existing issues by answering questions, creating reprexes, or providing feedback on proposed changes.

-   Contributing code or documentation via pull requests.

Anyone can become a contributor: there is no expectation of commitment to the project, no required set of skills, and no selection process.
We do not maintain an explicit list of contributors but acknowledge them in blog posts using `usethis::use_tidy_thanks()`, which gathers contribution data from the GitHub API.

### Authors

Contributors who have made significant and sustained contributions can be invited to become authors.
Authors are collectively responsible for day-to-day development of the package, including responding to issues and reviewing pull requests.
An author possesses two special powers:

> -   They have **write** access on GitHub so they can triage issues, request review on PRs, and merge them.

Reviewer: When you say "added to the tidyverse repo as whole", what exactly does that mean? I suspect means "add as member of the tidyverse organization".

> -   They have **write** access on GitHub so they can triage issues, request review on PRs, and merge them.

Author: Yes, I mean: should we create a "author" team for the tidyverse org, and add all author to that team?

> -   They have **write** access on GitHub so they can triage issues, request review on PRs, and merge them.

Author: Resolved to create an author team in the meeting yesterday.

-   Set and clearly communicates the strategic objectives of the project.
-   Oversee CRAN releases.
-   On-board new authors.
-   Mediate conflicts amongst authors.
-   Enforce the code of conduct.
-   Recruit a new maintainer when ready to retire from the project.

The maintainer is listed in `Authors@R`.
They must list their email address and be identified by the "cre" (creator) role.
They have **admin** access on Github, allowing them to add new authors when needed.

Maintainer turnover is slow and we have not yet developed a process for it, but we'd generally expect a maintainer to be a long-standing author.

### Common tasks

#### Author invitation

To on-board a new author, the maintainer first emails all other authors, checking for any major concerns.
If all authors are agreeable (or don't response within 7 days), the maintainer then sends the following email:

> Hi {name},
>
> In recognition of your significant contributions to {package}, would you be interested in becoming a co-author of the package?
> This means that you'll be acknowledged in [Authors\@R](mailto:Authors@R){.email} and given write permission on GitHub.

> > Write permission gives you the power to change dbplyr directly (which is fine for smaller fixes) but I'd appreciate it if you'd continue to send major stuff through the pull request process.

Reviewer 2: dbplyr should be changed to {package}

Author: Thanks!

---
output: github_document
---

# Tidyup 4: governance model

**Champion**: Hadley Wickham  
**Co-champion**: Tracy Teal  

> **Status**: Proposa
> l

Reviewer 3: ```suggestion
**Status**: Proposal
```

Author: This is a bug with the visual editor that I'm still trying to reprex for JJ 😞

## Abstract

Our open source packages currently don't have a well defined governance model.
We do have a basic set of assumptions under which we operate, but we have not previously made them precise.

> This tidyup lays out a model that strikes a balance between the benevolent dictator and rough consensus models, and defines the relationships between four key roles: a large community of package **users,** a smaller pool of GitHub **contributors**, a team of **authors**, and a single **maintainer**.

Reviewer 3: ```suggestion
This tidyup lays out a model that strikes a balance between the [benevolent dictator](https://en.wikipedia.org/wiki/Benevolent_dictatorship) and [rough consensus](https://en.wikipedia.org/wiki/Rough_consensus) models, and defines the relationships between four key roles: a large community of package **users,** a smaller pool of GitHub **contributors**, a team of **authors**, and a single **maintainer**.
```

Reviewer 3: Oh I see now that you link these below

## Motivation

We currently have one package (ggplot2) with a [formal governance model](https://github.com/tidyverse/ggplot2/blob/master/GOVERNANCE.md), and an informal, undocumented model that we've used for other packages (e.g. dtplyr, dbplyr).
The goal of this tidyup is to come up with holistic, flexible model, that can become our default governance model going forward.

> This is part of a general movement to get better at defining and documenting our processes to so that there's a clear path from package user to package developer.

Reviewer 4: ```suggestion
This is part of a general movement to get better at defining and documenting our processes so that there's a clear path from package user to package developer.
```

Author: Thanks! (merged manually)

We blend these two governance models because we want to involve the community in decision making as much as possible, while recognising that excellent user experience often requires consistent design guided by a single voice.

### Users

People who use the package are the most important members of the community; without these users, this project would have no purpose.
Users are encouraged to participate in the life of the project and the community as much as possible.
Common user activities include (but are not limited to):

-   Evangelising about the project.
-   Asking and answering questions on community forums.
-   Providing moral support (a "thank you" goes a long way).

### Contributors

Users who continue to engage with the project and its community will often find themselves becoming more and more involved.
Such users may then go on to become **contributors** by interacting with the project on GitHub.
Contributors:

-   Report bugs and suggest improvements by creating new issues.

> -   Improving existing issues by answering questions, creating reprexes, or providing feedback on proposed changes.

Reviewer 5: ```suggestion
-   Improve existing issues by answering questions, creating reprexes, or providing feedback on proposed changes.
```

> -   Contributing code or documentation via pull requests.

Reviewer 5: ```suggestion
-   Contribute code or documentation via pull requests.
```

Anyone can become a contributor: there is no expectation of commitment to the project, no required set of skills, and no selection process.

> The only requirement is to follow the code of conduct.

Reviewer 5: Do we have a link to a standard code of conduct we use?

Reviewer 6: Yes, this is the standard one. https://github.com/tidyverse/ggplot2/blob/master/CODE_OF_CONDUCT.md Working on some updates to it too.

Author: I'll leave the CoC generic in this doc, and we can fill in later.

Packages don't maintain an explicit list of contributors but acknowledge them in blog posts using data from GitHub aggregated by `usethis::use_tidy_thanks()`.

### Authors

Contributors who have made significant and sustained contributions can be invited to become authors.
Authors are collectively responsible for day-to-day development of the package, including responding to issues and reviewing pull requests.
An author possesses two special powers:

-   They have **write** access on GitHub so they can triage issues, request review on PRs, and merge them.

-   They are listed in `Authors@R` so they receive credit when others cite the package.

Authors are expected to follow our standard processes, such as:

-   **Code contribution**: code is usually contributed via PR, even for authors who could push directly.

>     Particularly high-stakes project may want to protect the main branch and require "request reviews before merging".

Reviewer: Has anyone actually done this? I wonder if it would make normal development frustratingly slow for small stuff. But maybe maintainer has a permission level that allows them to commit straight to a protected branch.

>     Particularly high-stakes project may want to protect the main branch and require "request reviews before merging".

Reviewer 3: Yes I have done it on a number of repos. Maintainers can still commit to the main branch, and merge PRs without extra approval.

>     Particularly high-stakes project may want to protect the main branch and require "request reviews before merging".

Reviewer: Ah, ok then that removes my hesitance.

-   **Communication**: authors are involved in most of the interactions with contributors and thus need to set a welcoming and inclusive tone for the project.

-   **PR review**: all pull requests should be reviewed by at least one other author.
    In general, there is no expectation that PRs contain clean commit histories, but it's appreciated where possible.

-   **Backward compatibility**: any backward incompatible changes (i.e. changes that cause reverse dependencies to fail `R CMD check` or are likely to cause problems in user code) must be approved by the maintainer.

>     Significant backward incompatible changes need to be accompanied with a plan for how they will be communicated to the community.

Reviewer: It feels like the lifecycle system/package should come up here. Is it intentional to *not* be quite that specific?

Author: Right, this is just a high-level list of important stuff. It might go away altogether in favour of the longer process doc.

-   **CRAN releases**: package releases are made on an as-needed basis, and increment either the major, minor, or patch version depending on the scope of the release.
    The process itself is defined by `usethis::use_release_issue()`.

-   **Decision making:** when a package has multiple authors, where possible decisions are made using rough consensus amongst the authors.
    If consensus is hard to reach or taking too long, the maintainer will make a decision.

(We expect to flesh these processes out in the coming months.)

Authors are recruited from contributors.
An invitation to join the authors can be extended to anyone who has made significant and sustained contributions, and has acted in accordance with the code of conduct.
Any existing author can propose a contributor be invited by emailing the maintainer.

### Maintainer

A maintainer is the author with primary responsibility for the project.
As well as the responsibilities of an author, they also:

-   Set and clearly communicates the strategic objectives of the project.
-   Oversee CRAN releases.
-   On-board new authors.

> -   Mediate conflicts amongst authors.

Reviewer 5: A sub-bullet of this is knowing when to use GitHub's built in "lock conversations" features, either temporarily or permanently, when arguments on an issue get out of hand

Author: Good idea, I'll think about the wording.

Author: Tweaked above bullet to "De-escalate conflict amongst authors and/or users.". Hopefully that captures the spirit, without getting too caught up in the details.

The maintainers of most of RStudio's open source repositories are RStudio employees.
This is not a requirement but a recognition of the tension between making development open to all and ensuring that users can trust that a package will be maintained in the long term (i.e. 10+ years), which typically requires the maintainer be explicitly remunerated for their work.
Where the maintainer is not an employee of RStudio, we ask for the "right of first refusal" — if the maintainer wants to stop maintaining the package (for whatever reason) they first offer it back to RStudio.
In the future, we hope to find other ways of financial supporting maintainers apart from full-time employment.

### Common tasks

#### Author invitation

To on-board a new author, the maintainer looks for rough consensus amongst authors (which would include no objections within 7 days).
The maintainer then sends the following email:

> Hi {name},
>
> In recognition of your significant contributions to {package}, would you be interested in becoming a co-author of the package?
> This means that you'll be acknowledged in `Authors@R` and given write permission on GitHub.

> > Write permission gives you the power to change {package} directly (which is fine for smaller fixes) but I'd appreciate it if you'd continue to send major stuff through the pull request process.

Reviewer 5: ```suggestion
> Write permission gives you the power to change {package} directly (which is fine for smaller fixes), but I'd appreciate it if you'd continue to submit major changes through the pull request process.
```

>
> If you accept, can you please prepare a PR that:
>
> -   Adds your info to `Authors@R`

> > -   Tweaks `_pkgdown.yml` if you want to link somewhere from the pkgdown site

Reviewer 3: ```suggestion
> -   Tweaks `_pkgdown.yml` if you want your name to link to somewhere from the pkgdown site
```

> -   Advertises the change in NEWS.md
> -   Re-builds the documentation to get updated package docs
>

> > I'll then add you as admin, and approve the PR, then you can squash-merge it.

Reviewer: Is it `admin`? Feels somewhat contradictory or at least inconsistent with terms encountered earlier (maintainer ~ admin, author ~ write).

https://docs.github.com/en/organizations/managing-access-to-your-organizations-repositories/repository-permission-levels-for-an-organization#permission-levels-for-repositories-owned-by-an-organization

Reviewer 3: admin sounded weird to me as well.

I think the GitHub terminology would be something like 'collaborator with write access'

Author: Are you suggesting we give authors maintain access instead of admin?

<img width="365" alt="Screenshot 2021-07-27 at 5 25 21 AM" src="https://user-images.githubusercontent.com/4196/127032281-e55c507c-c67e-48a0-b324-51dc9f544d83.png">

Reviewer 3: I was assuming we would give them write? Why would authors need admin?

Author: Ooops I wrote "write" elsewhere so I was clearly just confused here.

> This would be our workflow going forward.
> (You'll also be able to request reviews from me and other authors as needed).
>
> Thanks for all your work on {package}!
>
> {your_name}

(After this tidyup is approved, the template will become a usethis function.)

## Open issues

> -   How can we communicate with authors? Do we add them to a team? Do we add them as members of the repo? Do we add them to a slack channel?

Reviewer 3: Team discussions could potentially work as well if people would prefer that https://docs.github.com/en/organizations/collaborating-with-your-team/creating-a-team-discussion

Though I do think having a slack channel with all the authors across our repos would probably be the best bet.

Reviewer: I worry about adding a new communication channel (team discussions) that we're not already using.

Author: Yeah, me too. I'll make a note of this concern in the writeup.

Where the maintainer is not an employee of RStudio, we ask for the "right of first refusal" — if the maintainer wants to stop maintaining the package (for whatever reason) they first offer it back to RStudio.
In the future, we hope to find other ways of financial supporting maintainers apart from full-time employment.

### Process

#### Invite author

To on-board a new author, the maintainer looks for rough consensus amongst authors by emailing them (since this needs to be done in private).
(one signal of rough consensus would be no objections within 7 days).
The maintainer then sends the following email:

> Hi {name},
>
> In recognition of your significant contributions to {package}, would you be interested in becoming a package author?
> You can read about the rights and responsibilities of a package author at ..., but in short, being an author means that you'll be acknowledged in `Authors@R` and given write permission on GitHub.

> > You'll continue to use PRs to contribute code, but merge your own PRs once they've been reviewed, and you can now review PRs from others.

Reviewer: Oddly, I think basically anyone can review PRs. It's *requesting* review that is a privilege that requires higher-than-read permission. Along with other stuff like assigning, labelling, etc.

Reviewer: Anyhow, maybe you just mean that morally an author should feel cool about reviewing, whereas it's more of a grey area for non-authors.

Author: Yeah, exactly.
